website/src/docs/hotchocolate/execution-engine/field-middleware.md

```diff
@@ -0,0 +1,291 @@
+    {
+        _next = next;
+    }
+
+    // this method must be called InvokeAsync or Invoke
+    public async Task InvokeAsync(IMiddlewareContext context)
+    {
+        // Code up here is executed before the following middleware
+        // and the actual field resolver
+
+        // This invokes the next middleware
+        // or if we are at the last middleware the field resolver
+        await _next(context);
+
+        // Code down here is executed after all later middleware
+        // and the actual field resolver has finished executing
+    }
+}
+```
+
+If you need to access services you can either inject them via the constructor, if they are singleton, or as an argument of the `InvokeAsync` method, if they have a scoped or transient lifetime.
+
+```csharp
+public class MyMiddleware
+{
+    private readonly FieldDelegate _next;
+    private readonly IMySingletonService _singletonService;
+
+    public MyMiddleware(FieldDelegate next, IMySingletonService singletonService)
+    {
+        _next = next;
+        _singletonService = singletonService;
+    }
+
+    public async Task InvokeAsync(IMiddlewareContext context,
+        IMyScopedService scopedService)
+    {
+        // Omitted code for brevity
+    }
+}
+```
+
+The ability to add additional arguments to the `InvokeAsync` method is the reason why there isn't a contract like an interface or a base class for field middleware.
+
+### Usage
+
+Now that you've defined the middleware as a class we need to still apply it to a field.
+
+```csharp
+public class QueryType : ObjectType
+{
+    protected override void Configure(IObjectTypeDescriptor descriptor)
+    {
+        descriptor
+            .Field("example")
+            .Use<MyMiddleware>()
+            .Resolve(context =>
+            {
+                // Omitted for brevity
+            });
+    }
+}
+```
+
+While an extension method like `UseMyMiddleware` on the `IObjectFieldDescriptor` doesn't make as much sense for `Use<MyMiddleware>` in contrast to the middleware delegate, we still recommend creating one as shown [here](#reusing-the-middleware-delegate). The reason being that you can make changes to this middleware more easily in the future without potentially having to change all places this middleware is being used in.
+
+If you need to pass an additional custom argument to the middleware you can do so using the factory overload of the `Use` method.
+
+```csharp
+descriptor
+    .Field("example")
+    .Use((provider, next) => new MyMiddleware(next, "custom",
+        provider.GetRequiredService<FooBar>()));
+```
+
+# Usage as an attribute
+
+Up until now we have only worked with Code-first APIs to create the field middleware. What if you want to apply your middleware to a field resolver defined using the Annotation-based approach?
+
+You can create a new attribute inheriting from `ObjectFieldDescriptorAttribute` and call or create your middleware inside of the `OnConfigure` method.
+

```

Comment:
```suggestion

> Note: Attribute order is not guaranteed in C# which is why we in the case of middleware attributes use the `CallerLineNumberAttribute` to inject at compile time the C# line number. The line number is used as an order. We do not recommend inheriting middleware attributes from a base field since this can lead to confusion about ordering. Look at the example below how we infer the order. When inheriting from middleware attributes always pass through the order argument. Further, indicate with the `Use` verb that your attribute is a middleware attribute.

```